packages/runtime/src/runtimes/react/controls/rich-text-v2.tsx

```diff
@@ -0,0 +1,171 @@
+import {
+  FocusEvent,
+  KeyboardEvent,
+  MutableRefObject,
+  ReactNode,
+  useCallback,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react'
+import { Descendant, Editor, createEditor, Range as SlateRange } from 'slate'
+import isHotkey from 'is-hotkey'
+import { withHistory, HistoryEditor } from 'slate-history'
+import {
+  withReact,
+  ReactEditor,
+  RenderElementProps,
+  RenderLeafProps,
+  Slate,
+  Editable,
+} from 'slate-react'
+
+import {
+  BlockType,
+  RichTextV2Control,
+  RichTextV2ControlData,
+  RichTextV2ControlDefinition,
+} from '../../../controls'
+import { useBuilderEditMode } from '..'
+import { BuilderEditMode } from '../../../state/modules/builder-edit-mode'
+import { pollBoxModel } from '../poll-box-model'
+import { useIsomorphicLayoutEffect } from '../../../components/hooks/useIsomorphicLayoutEffect'
+
+export type RichTextV2ControlValue = ReactNode
+
+export type Descriptors = { text?: RichTextV2ControlDefinition }
+
+const defaultText: Descendant[] = [{ type: BlockType.Text, children: [{ text: '' }] }]
+
+/**
+ * Clicking outside of the host blurs our `<Editable />`.
+ * `<Editable />` only updates the DOM's selection to match slate when it is focused.
+ * In the case of a panel being clicked this hook updates the DOM selection to match slate.
+ */
+export function useSyncDOMSelection(editor: Editor, isEnabled: MutableRefObject<boolean>) {
+  useIsomorphicLayoutEffect(() => {
+    if (!isEnabled.current || editor.selection == null || ReactEditor.isFocused(editor)) return
+    try {
+      const root = ReactEditor.findDocumentOrShadowRoot(editor) as Document
+      const domSelection = root.getSelection()
+      const newDomRange: Range | null = ReactEditor.toDOMRange(editor, editor.selection)
+
+      if (newDomRange) {
+        if (SlateRange.isBackward(editor.selection!)) {
+          domSelection?.setBaseAndExtent(
+            newDomRange.endContainer,
+            newDomRange.endOffset,
+            newDomRange.startContainer,
+            newDomRange.startOffset,
+          )
+        } else {
+          domSelection?.setBaseAndExtent(
+            newDomRange.startContainer,
+            newDomRange.startOffset,
+            newDomRange.endContainer,
+            newDomRange.endOffset,
+          )
+        }
+      } else {
+        domSelection?.removeAllRanges()
+      }
+    } catch (e) {
+      console.error(e)
+    }
+  })
+}
+
+export function useRichTextV2(data: RichTextV2ControlData, control: RichTextV2Control | null) {
```

Comment:
Minor point, but we might want to actually create a `RichTextV2` component in this file instead of embedding what looks like a component into this hook. Then, this hook can return `<RichTextV2 />` passing the relevant props that are outside the scope of the component, like getting the control instance, communicating with the builder control, etc.